### ISSUE TITLE
Make sense to prime shoebox from express?

### ISSUE THREAD

**Issue by kellyselden** [ember-fastboot/fastboot#82]
It would be terribly useful if I could start adding to the shoebox in express before my app starts rendering in fastboot. Possible API could be:

```js
let app = new FastBoot({
  distPath: 'path/to/dist',
  shoebox: {
    // ...
  }
});
```

or

```js
app.visit('/photos', {
  shoebox: {
    // ...
  }
}).then(// ...
```

Does this make sense? Is there an existing alternative way to do this?

**Comment by jasonmit** [comment]
There is another PR opened for adding metadata/state to the fastboot service before the app is instantiated.

Perhaps both efforts should be consolidated into a single API.

**Comment by kellyselden** [comment]
Ah yes I see it #75. Didn't come up when I searched `shoebox`. I think using the existing shoebox instead of a new metadata api will make it more appealing.

**Comment by jasonmit** [comment]
https://github.com/ember-fastboot/ember-cli-fastboot/pull/245

Is that PR mentioned above

**Comment by jasonmit** [comment]
Agreed, just mentioning so both are not worked on in favor of a single solution.

**Comment by kellyselden** [comment]
And worth noting if we used the shoebox, no `ember-cli-fastboot` changes would be necessary.

**Comment by kellyselden** [comment]
Alright I'm going to fork and start working on this, because I think it's required for me to continue porting my express app to run fastboot.

@jasonmit Is `visit` the better place to start the shoebox over `new FastBoot`? Are you ever going to change shoebox values for different requests?

**Comment by jasonmit** [comment]
This was Tom's feedback when I first proposed this idea over slack.  I think for us though we needed access before the visit took place.

Cc @thoov who might have more context; I'm still on vacation without access to my project where it was needed

**Comment by kellyselden** [comment]
> I think for us though we needed access before the visit took place.

Before the visit, you're not in Ember. So I'm confused how you needed it before then?

**Comment by thoov** [comment]
@kellyselden What type of info do you want to pre-seed with? We came up with passing some data into the visit via metadata [#75](https://github.com/ember-fastboot/fastboot/pull/75) and then getting it back out [#245](https://github.com/ember-fastboot/ember-cli-fastboot/pull/245) via: `fastbootService.get('metadata')`.

**Comment by jasonmit** [comment]
> Before the visit, you're not in Ember. So I'm confused how you needed it before then?

The state needed to be available within Instance initializers and before the router is setup.  Unsure, without looking closely, if that can be accomplished through `visit` versus through the constructor.  If `visit`, then I'm all in favor of the API you proposed.

The use case there is registering routes specific to the user, preloading the app with translations, and a few other things dependent on request context.

@thoov I think it would be a good idea if instead of having two places to hang state that it just leverage the shoebox API.

**Comment by kellyselden** [comment]
Closed in favor of #75, where you can put stuff in `metadata`, then convert to shoebox in a fastboot initializer.
